diff -up mesa-1/src/mesa/gl.pc.in.jx mesa-1/src/mesa/gl.pc.in
--- mesa-1/src/mesa/gl.pc.in.jx	2018-08-17 12:08:20.764877163 +0200
+++ mesa-1/src/mesa/gl.pc.in	2018-08-17 12:08:41.381604571 +0200
@@ -7,7 +7,7 @@ Name: gl
 Description: Mesa OpenGL library
 Requires.private: @GL_PC_REQ_PRIV@
 Version: @PACKAGE_VERSION@
-Libs: -L${libdir} -l@GL_PKGCONF_LIB@
+Libs: -L${libdir} -lGL
 Libs.private: @GL_PC_LIB_PRIV@
 Cflags: -I${includedir} @GL_PC_CFLAGS@
 glx_tls: @GLX_TLS@
